src/cart/crt_context.c

```diff
@@ -948,6 +948,17 @@ crt_context_req_untrack(struct crt_rpc_priv *rpc_priv)
 	D_INIT_LIST_HEAD(&submit_list);
 
 	D_MUTEX_LOCK(&epi->epi_mutex);
+
+	/* Prevent simultaneous untrack from progress thread and
+	 * main rpc execution thread.
+	 */
+	if (rpc_priv->crp_in_binheap == 0) {
```

Comment:
if crt_rpc.c:542 executes first, we still get double untrack, because there the rpc isn't taken out from the binheap.

Reply:
good point. after discussion changed hg_addr lookup failures during lc_addr_insert to mark rpc state as unreachable.

Reply:
> good point. after discussion changed hg_addr lookup failures during lc_addr_insert to mark rpc state as unreachable.

You did change something, or you will change something?  It's not clear what you mean here or if you've changed the PR since.

Reply:
I have uploaded patch #2 to address yulus find. by marking rpc as unreachable in such situations it now will remove rpc from binheap on those errors